### ISSUE TITLE
Composition cleaning QA Pt 1. (Main Index page)

### ISSUE THREAD

**Issue by Sakshi-Handa** [HumanExposure/factotum#2049]
As a product owner
I would like to view all Composition Cleaning Scripts that contain manually extracted documents
So that I can determine the status of Composition Cleaning QA and access the QA workflow.

We would like to develop a QA workflow for checking the quality of composition cleaning scripts. 
After a composition document is extracted, we perform an additional curation step of 'cleaning' the raw composition data to harmonize the values and units into weight fractions. So we need a way to check the quality of the cleaning step, by comparing the reported composition values to the cleaned weight fractions. 
This new QA workflow will be 'script-based', similar to the Composition Extraction Script QA
https://factotum.epa.gov/qa/extractionscript/

The QA Group will consist of a random subset of 20% of the extracted text documents associated with a cleaning script. If there are less than 50 documents, the QA group includes all documents. 
(This is similar to the logic implemented for Composition Extraction Script QA, where we QA up to 100 documents, or 20% of the documents for large groups.)

Definition of Done:
· Create a new page for QA Composition Cleaning Script, similar to the 'Composition Extraction Script QA'
· Add a dropdown link to the new QA page from the 'QA' tab on the main nav bar
· Create a QA Group associated with each script, that contains a randomly selected 20% of documents associated with a given cleaning scrip for groups of more than 50 documents. If there are less than 50 documents, the QA Group contains all documents.
· Create a table with the following columns:
	• Cleaning script (link to the cleaning script in Github)
	• Script updated (date the script was last updated)
	• Total Documents ( number of documents associated with the script)
	• QA Group Documents (number of documents to complete QA approval)
	• Percent of QA Group Checked (% of documents in QA Group that have completed QA)
	• QA Summary (QA Summary Page will be created in pt. 3)
	• QA Status (QA Group page will be created in pt. 2)

**Comment by Sakshi-Handa** [comment]
Instead of 'QA approved', which is currently used for Extraction QA,
add necessary fields for Cleaning QA at the document level.

**Comment by mwfrost** [comment]
@Sakshi-Handa 
I think this issue exposes a data modeling error. The cleaning script is assigned as the `script_id` attribute on the `ExtractedComposition` subclass, so it's stored at the chemical level. Back in 2019, we were instructed to store the attribute there, rather than at the data document level. 

When the data document page is rendered, however, it shows a single link to the cleaning script, which gives the impression that the cleaning script is a document-level attribute. Behind the scenes that script is just selected as the cleaning script assigned to an arbitrarily first-chosen chemical under a Composition data document. 

This looks to me like a workaround, not the way it's meant to behave. The cleaning script, from what I can tell, should be stored at the document (or `ExtractedText`) level. I queried the database and found zero instances of a document where more than one cleaning script was used on its chemical records. 

Can we use this ticket to pay down that technical debt by migrating the cleaning script up from the chemical level to the document level?

**Comment by Sakshi-Handa** [comment]
Yes, we can do this first. It makes more sense for the script to be stored at the document level. I don't think we would ever have more than 1 cleaning script associated with a document. 

I wonder if this was designed this way because of our old functional use cleaning workflow, which would have required uploading cleaning scripts. Then maybe it would have made more sense to associate the cleaning script with the chemical, so you can have multiple scripts for a document.

**Comment by mwfrost** [comment]
@Sakshi-Handa 
Are the cleaned composition fields subject to the same QA tracking that the other extracted chemical fields are? If a user changes an extracted field in a data document, they have to add a QA Note before it can be approved. Should the same constraints apply to the cleaned fields and the cleaning QA process?

**Comment by mwfrost** [comment]
And if a document enters this "unapprovable" state because of edits to its extracted data, is it also unapprovable for cleaning purposes? Or are the two QA tracks totally separate?

**Comment by Sakshi-Handa** [comment]
> **Are the cleaned composition fields subject to the same QA tracking that the other extracted chemical fields are?** 

Do you mean if they are tracked in the audit log? Yes they are tracked by min/max/central 'wf_analysis'. 

>**If a user changes an extracted field in a data document, they have to add a QA Note before it can be approved. Should the same constraints apply to the cleaned fields and the cleaning QA process?**

We currently don't have a method for removing/editing individual cleaned composition fields in a chemical record. I think we don't want to edit individual cleaned composition records. We were planning to handle that by script ( #2053 ). 
If someone QA checks a 'Composition- Cleaning Script' document and finds an error, we wouldn't want them to fix it in that single record, or single document. We would expect that 'systematic' error to show up in all documents related to that cleaning script. So we want to fix the cleaning script and re-upload the cleaned data. 
In that case, maybe we need to require the QA Note when a curator 'rejects' the document. To say why the document does not pass the QA inspection. 

>**And if a document enters this "unapprovable" state because of edits to its extracted data, is it also unapprovable for cleaning purposes? Or are the two QA tracks totally separate?**

I think the two QA tracks are totally separate. Because ideally, we would have to approve the "Composition -Extraction Script" QA first, as a prerequisite for cleaning the composition data. If there were any issues with the extraction script, they would be addressed and corrected before we clean the data. Then once we upload cleaned composition data, we can check the quality of the cleaned weight fractions with the "Composition- Cleaning Script" QA. 
Maybe that was a confusing explanation. Basically our workflow is like this:
1. Extract data from Composition documents (by script/manually)
2. QA the Extracted Text (If Extraction QA is acceptable quality, proceed to 3.Composition Cleaning. Else Redo Extraction Script)
3. Clean Composition Data
4. QA the Cleaned Composition Data (if Cleaning QA is acceptable, proceed ... Else Redo Cleaning Script)

**Comment by mwfrost** [comment]
@Sakshi-Handa Thanks!
